pkg/volumes/volumes.go

```diff
@@ -0,0 +1,126 @@
+// structure
+func NewStepVolumeMount(name string, readOnly bool) StepVolumeMount {
+	return StepVolumeMount{
+		Name:     name,
+		ReadOnly: readOnly,
+	}
+}
+
+// TaskSpecVolumes creates a list of Volumes for the generated TaskSpec object and
+// checks for some erroneous situations around volumes and volume mounts
+func TaskSpecVolumes(
+	existingVolumeMounts []StepVolumeMount,
+	strategyVolumes []buildv1alpha1.BuildVolume,
+	buildVolumes []buildv1alpha1.BuildVolume,
+	buildrunVolumes []buildv1alpha1.BuildVolume,
+) ([]corev1.Volume, error) {
+	res := []corev1.Volume{}
+
+	// first we merge build volumes into the strategy ones, next we merge
+	// build run volumes into result of the previous merge.
+	// eventual list of volumes will be added to the generated TaskSpec object
+	volumes, err := MergeBuildVolumes(strategyVolumes, buildVolumes)
+	if err != nil {
+		return nil, err
+	}
+	volumes, err = MergeBuildVolumes(volumes, buildrunVolumes)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range volumes {
+		volumeMountExists := false
+		for _, volumeMount := range existingVolumeMounts {
+			if volumeMount.Name == v.Name {
+				volumeMountExists = true
+
+				// In case volume mount is not read only and
+				// volume type is either secret or config map,
+				// build should not be run, because this situation may lead
+				// to errors
+				if !volumeMount.ReadOnly &&
+					(v.VolumeSource.ConfigMap != nil ||
+						v.VolumeSource.Secret != nil) {
+					return nil,
+						fmt.Errorf("Volume Mount %q must be read only", v.Name)
+				}
+			}
+		}
+
+		if !volumeMountExists && v.Optional != nil && !*v.Optional {
+			return nil, fmt.Errorf("Volume %q is not optional and does not have corresponding Volume Mount", v.Name)
+		}
+
+		taskRunVolume := corev1.Volume{
+			Name:         v.Name,
+			VolumeSource: v.VolumeSource,
+		}
+
+		res = append(res, taskRunVolume)
+	}
+
+	return res, nil
+}
+
+// MergeBuildVolumes merges Build Volumes from one list into the other. It only allows to merge those that have property
+// Overridable set to true. In case it is empty or false, it is not allowed to be overridden, so Volume cannot be merged
+// Merging in this context means copying the VolumeSource from one object to the other.
+func MergeBuildVolumes(into []buildv1alpha1.BuildVolume, new []buildv1alpha1.BuildVolume) ([]buildv1alpha1.BuildVolume, error) {
+	if len(new) == 0 && len(into) == 0 {
+		return []buildv1alpha1.BuildVolume{}, nil
+	}
+	if len(new) == 0 {
+		return into, nil
+	}
+
+	var result []buildv1alpha1.BuildVolume
+	var errors []error
+
+	for _, vol := range into {
+		result = append(result, *vol.DeepCopy())
+	}
+
+	for _, merger := range new {
+		found := false
+		for i, original := range into {
+			if original.Name == merger.Name {
+				if original.Overridable != nil && !*original.Overridable {
+					errors = append(errors, fmt.Errorf("Cannot override BuildVolume %q", original.Name))
```

Comment:
These validations should also happen in the Build registration. Scenario-wise, this is similar to the validation of paramValues. If a Build specify a volume that does not exist on the build strategy, or is not overridable, then the Build should not get registered. We should have a set of reasons for the different error scenarios (volume not defined, volume not overridable).